### ISSUE TITLE
[Bug]:  {\"message\":\"ERROR: portal snapshots (0) did not account for all active snapshots (1) (SQLSTATE XX000)\"} when refreshing a continuous aggregate via a trigger

### ISSUE THREAD

**Issue by choec0130** [timescale/timescaledb#6533]
### What type of bug is this?

Unexpected error

### What subsystems and features are affected?

Continuous aggregate

### What happened?

I am trying to call refresh my continuous aggregate via a trigger and get this error 
`{\"message\":\"ERROR: portal snapshots (0) did not account for all active snapshots (1) (SQLSTATE XX000)\"}`

here's my trigger 
`create trigger refresh_aggregates
    after update or insert or delete
    on pl
execute function refresh_pl_continuous_aggregates();`

and my trigger function 
`create function refresh_pl_continuous_aggregates() returns trigger as
$$
BEGIN
    CALL refresh_continuous_aggregate('aggregates_by_pl_by_day', '2024-01-01', INTERVAL '2 day');
END;
$$
LANGUAGE 'plpgsql';`

continuous aggregate
`CREATE MATERIALIZED VIEW aggregates_by_pl_by_day
            WITH (timescaledb.continuous) AS
SELECT time_bucket('1 day' , modified_date) as day_bucket,
       sl.pl_id,
       resource_type,
       taxonomy,
FROM activities
    JOIN s_licenses sl ON sl.s_id = activities.s_id
GROUP BY 1, sl.pl_id, resource_type, taxonomy;`

### TimescaleDB version affected

2.13.0

### PostgreSQL version used

15.5

### What operating system did you use?

Ubuntu 15.5-1.pgdg22.04+1 on x86_64-pc-linux-gnu

### What installation method did you use?

Not applicable

### What platform did you run on?

Timescale Cloud

### Relevant log output and stack trace

```bash
{\"message\":\"ERROR: portal snapshots (0) did not account for all active snapshots (1) (SQLSTATE XX000)\"}
```

### How can we reproduce the bug?

```bash
1. Create an "after" trigger that manually refreshes a continuous aggregate using `refresh_continuous_aggregate`
2. Set off the trigger.
3. Notice that the continuous aggregate doesn't refresh and instead the error is produced.
```

**Comment by konskov** [comment]
Hello @choec0130 , thank you for reaching out. 
Sharing a minimal example to reproduce, will look further into the issue.
```
CREATE TABLE metrics (
    time timestamptz NOT NULL,
    device_id int,
    device_id_peer int,
    v0 int,
    v1 int,
    v2 float,
    v3 float
);

SELECT create_hypertable('metrics', 'time');

INSERT INTO metrics (time, device_id, device_id_peer, v0, v1, v2, v3)
SELECT time,
    device_id,
    0,
    device_id + 1,
    device_id + 2,
    0.5,
    NULL
FROM generate_series('2024-01-01 0:00:00+0'::timestamptz, '2000-01-04 23:55:00+0', '20m') gtime (time),
    generate_series(1, 2, 1) gdevice (device_id);

CREATE MATERIALIZED VIEW metrics_cagg WITH (timescaledb.continuous,
  timescaledb.materialized_only = true)
AS
SELECT time_bucket('1 day', time) as dayb, device_id,
       sum(v0), avg(v3)
FROM metrics
GROUP BY 1, 2
WITH NO DATA;

CREATE FUNCTION refresh_pl_continuous_aggregates() RETURNS TRIGGER AS $$ BEGIN CALL refresh_continuous_aggregate(
    'metrics_cagg',
    '2024-01-01',
    INTERVAL '2 day'
);
END;
$$ LANGUAGE 'plpgsql';

CREATE TRIGGER refresh_aggregates AFTER UPDATE OR INSERT OR DELETE ON metrics
EXECUTE FUNCTION refresh_pl_continuous_aggregates();
```
As a side note, you can use realtime aggregation in your caggs (by specifying materialized_only = false) and adjust the schedule interval of a refresh policy to run as frequently as you need, in order to get up-to-date values when querying your continuous aggregate, instead of using a trigger like this.

**Comment by nikkhils** [comment]
@konskov looks like we should disallow creation of triggers on CAggs based on the above.

**Comment by choec0130** [comment]
@konskov Thank you! I have realtime aggregation on but I'm actually using the trigger in combination with invalidating the continuous aggregate because I'm having an update on a joined table and not the hypertable itself, which is why I'm setting up a trigger. Basically the solution talked about here: https://github.com/timescale/timescaledb/issues/6353

Thanks so much for looking into this!

**Comment by aiman-saeed** [comment]
I faced this too on `timescale/timescaledb-ha:pg15.8-ts2.17.1` - whenever I prepare a dynamic statement.

For instance, this works:
`CALL refresh_continuous_aggregate('visual_ds', NULL, INTERVAL '1 week');`

But, this doesn't:
CALL refresh_continuous_aggregate('fl_visual_ds_view_l7', NULL, INTERVAL '1 week');
`EXECUTE format('CALL refresh_continuous_aggregate(%L, NULL, NULL)', _cagg.user_view_name);`

**Comment by staticlibs** [comment]
I think the current behaviour is correct - `refresh_continuous_aggregate` uses transactions internally, so cannot be called from functions or from dynamic SQL. But the error message reported could have been better. There is a transaction check in `refresh_continuous_aggregate` but it seems to be ineffective. I've filed #7566 PR to improve this check and report one of the following error messages:

 - `ERROR:  refresh_continuous_aggregate() cannot run inside a transaction block`
 - `ERROR:  refresh_continuous_aggregate() cannot be executed from a function`
 
instead of `ERROR: portal snapshots (N) did not account for all active snapshots (N+1)`.

**Comment by aiman-saeed** [comment]
@staticlibs @fabriziomello

I do hierarchical aggregation and I use a [procedure](https://ideia.me/customize-multiple-refreshes-continuous-aggregates) to refresh the most lower level (minutes) before proceeding to upper levels (hours, then days). Now, I also want to add a capacity to do a full clean refresh from scratch based of the raw hyper-table, but in short time periods to avoid running out of `max_locks_per_transaction` as a full refresh will lock multiple chunks.

In summary, I could really benefit from a refresh method that could be prepared dynamically in a procedure, this would enable me to pass in start/end times dynamically. I am looking forward to hear if there could be any workarounds! 

Thank-you!

**Comment by staticlibs** [comment]
@aiman-saeed 

Perhaps something like this can work for you, where the `SELECT` that returns refresh params can be constructed dynamically:

```plpgsql
DO $$
DECLARE
    rec record;
BEGIN
    FOR rec IN 
        EXECUTE $inner$
            SELECT * FROM (
                VALUES 
                    ('metrics_cagg_1', '2024-01-01'::timestamptz, INTERVAL '1 day'),
        	    ('metrics_cagg_2', '2024-01-02'::timestamptz, INTERVAL '2 day')
	    ) AS p (cagg_name, window_start, window_end)
        $inner$
    LOOP
        RAISE NOTICE 'params: id=%, name=%, value=%', rec.cagg_name, rec.window_start, rec.window_end;            
        CALL refresh_continuous_aggregate(rec.cagg_name, rec.window_start, rec.window_end);
    END LOOP;
END;
$$;
```
